File: python/Q621_task_scheduler.py
#
# @lc app=leetcode id=621 lang=python3
#
# [621] Task Scheduler
#

# @lc code=start
import heapq
from collections import deque


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        remaining: dict[str, int] = dict()
        for t in tasks:
            if t not in remaining:
                remaining[t] = 0

            remaining[t] += 1

        max_heap: list[int] = [-count for _, count in remaining.items()]
        heapq.heapify(max_heap)

        time: int = 0
        processing: deque[tuple[int, int]] = deque()

        while max_heap or processing:
            if max_heap:
                count = heapq.heappop(max_heap)

                if count + 1 < 0:
                    processing.append((count + 1, time + n))

            if processing and processing[0][1] == time:
                count, _ = processing.popleft()
                heapq.heappush(max_heap, count)

            time += 1

        return time


# @lc code=end

File: rust/src/leetcode/Q621_task_scheduler.rs
/*
 * @lc app=leetcode id=621 lang=rust
 *
 * [621] Task Scheduler
 */

// @lc code=start
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::collections::VecDeque;

impl Solution {
    pub fn least_interval(tasks: Vec<char>, n: i32) -> i32 {
        let mut task_map: HashMap<char, i32> = HashMap::new();

        for &c in tasks.iter() {
            *task_map.entry(c).or_default() += 1
        }

        let mut task_heap: BinaryHeap<i32> = BinaryHeap::new();
        for (_, &value) in task_map.iter() {
            task_heap.push(value);
        }

        let mut time: i32 = 0;
        let mut processing: VecDeque<(i32, i32)> = VecDeque::new();

        while !processing.is_empty() || !task_heap.is_empty() {
            if let Some(count) = task_heap.pop() {
                if count != 1 {
                    processing.push_back((time + n, count - 1));
                }
            }

            if let Some(&(ready_time, count)) = processing.front() {
                if ready_time == time {
                    processing.pop_front();
                    task_heap.push(count);
                }
            }

            time += 1;
        }

        time
    }
}
// @lc code=end
